server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala

```diff
@@ -108,15 +251,212 @@ case class EmailHeaders(headers: List[EmailHeader],
+          bodyStructure = bodyStructure,
+          textBody = bodyStructure.textBody,
+          htmlBody = bodyStructure.htmlBody,
+          attachments = bodyStructure.attachments,
+          bodyValues = bodyValues,
+          hasAttachment = HasAttachment(!firstMessage.getLoadedAttachments.isEmpty),
+          preview = preview),
+        specificHeaders = EmailHeaders.extractSpecificHeaders(request.properties)(mime4JMessage))
+    }
+  }
+
+  private def extractBodyValues(bodyStructure: EmailBodyPart, request: EmailGetRequest): Try[Map[PartId, EmailBodyValue]] = for {
+    textBodyValues <- extractBodyValues(bodyStructure.textBody, request, request.fetchTextBodyValues.exists(_.value))
+    htmlBodyValues <- extractBodyValues(bodyStructure.htmlBody, request, request.fetchHTMLBodyValues.exists(_.value))
+    allBodyValues <- extractBodyValues(bodyStructure.flatten, request, request.fetchAllBodyValues.exists(_.value))
+  } yield {
+    (textBodyValues ++ htmlBodyValues ++ allBodyValues)
+      .distinctBy(_._1)
+      .toMap
+  }
+
+  private def extractBodyValues(parts: List[EmailBodyPart], request: EmailGetRequest, shouldFetch: Boolean): Try[List[(PartId, EmailBodyValue)]] =
+    if (shouldFetch) {
+      parts
+        .map(part => part.bodyContent.map(bodyValue => bodyValue.map(b => (part.partId, b.truncate(request.maxBodyValueBytes)))))
+        .sequence
+        .map(list => list.flatten)
+    } else {
+      Success(Nil)
+    }
+}
+
+private class EmailMetadataViewReader @Inject()(messageIdManager: MessageIdManager,
+                                                metadataViewFactory: EmailMetadataViewFactory) extends EmailViewReader {
+  private val reader: GenericEmailViewReader = new GenericEmailViewReader(messageIdManager, FetchGroup.MINIMAL, metadataViewFactory)
+
+  override def read(ids: Seq[MessageId], request: EmailGetRequest, mailboxSession: MailboxSession): SFlux[EmailView] =
+    reader.read(ids, request, mailboxSession)
+}
+
+private class EmailHeaderViewReader @Inject()(messageIdManager: MessageIdManager,
+                                              headerViewFactory: EmailHeaderViewFactory) extends EmailViewReader {
+  private val reader: GenericEmailViewReader = new GenericEmailViewReader(messageIdManager, FetchGroup.HEADERS, headerViewFactory)
+
+  override def read(ids: Seq[MessageId], request: EmailGetRequest, mailboxSession: MailboxSession): SFlux[EmailView] =
+    reader.read(ids, request, mailboxSession)
+}
+
+private class EmailFullViewReader @Inject()(messageIdManager: MessageIdManager,
+                                            metadataViewFactory: EmailFullViewFactory) extends EmailViewReader {
+  private val reader: GenericEmailViewReader = new GenericEmailViewReader(messageIdManager, FetchGroup.FULL_CONTENT, metadataViewFactory)
+
+  override def read(ids: Seq[MessageId], request: EmailGetRequest, mailboxSession: MailboxSession): SFlux[EmailView] =
+    reader.read(ids, request, mailboxSession)
+}
```

Comment:
Wouldn't it be better to extract out the `Reader`s and `Factory`s? Start to be a lot of classes in Email file

Reply:
Well still it's about instanciating EmailView, I believe it still stands in a single file.